### PR TITLE
HTTP cookie: A better parser.

### PR DESCRIPTION
**Pull request by changxiaocui**
The previous parser had problems with "hello=str1=str2;" which name is "hello" and value is "str1=str2".

### CODE REVIEW

**Comment by hongzhidao** [comment]
Thank you for the patch.

1. There is an issue as you found.
2. I wonder if your code has refactoring in addition to fixing the issue. If yes, I suggest separating them into two commits. It would help us to review.  For example:
    ```
static nxt_http_name_value_t *
nxt_http_cookie(nxt_array_t *array, u_char *name, size_t name_length,
    u_char *start, u_char *end)
nxt_http_cookie(nxt_array_t *array, nxt_str_t *name, nxt_str_t *value)
```
It looks like a refactoring.

**Comment by hongzhidao** [comment]
btw, could you provide corresponding test cases?

**Comment by changxiaocui** [comment]
In fact, It's hard to fix these issues without refactoring. 

> Thank you for the patch.
> 
> 1. There is an issue as you found.
> 2. I wonder if your code has refactoring in addition to fixing the issue. If yes, I suggest separating them into two commits. It would help us to review.  For example:
> 
> ```
> static nxt_http_name_value_t *
> nxt_http_cookie(nxt_array_t *array, u_char *name, size_t name_length,
>     u_char *start, u_char *end)
> nxt_http_cookie(nxt_array_t *array, nxt_str_t *name, nxt_str_t *value)
> ```
> 
> It looks like a refactoring.

In fact, It's hard to fix these issues without refactoring.

**Comment by changxiaocui** [comment]
> btw, could you provide corresponding test cases?

It will take some time.

**Comment by hongzhidao** [comment]
Okay, could you add simple descriptions for what issues are fixed? Or adding tests in the test dir if you know how to do it.

**Comment by alejandro-colomar** [comment]
> In fact, It's hard to fix these issues without refactoring.

Maybe you can refactor without fixing the issues?  So you could refactor as the first commit; and then fix the issue.

**Comment by changxiaocui** [comment]
After trying, I decided to add descriptions instead of adding tests.

As per https://datatracker.ietf.org/doc/html/rfc6265#section-4.1.1
cookie-value = *cookie-octet / ( DQUOTE *cookie-octet DQUOTE )
cookie-octet = %x21 / %x23-2B / %x2D-3A / %x3C-5B / %x5D-7E

So '=' is a valid character in cookie-value, but the previous parser ignored this.
The code of the new parser is clear enough to read.

Below is some code when I tried to add tests. Most of the examples are from
the google search client.

`static struct {
nxt_str_t cookie;
nxt_str_t name;
nxt_str_t value;

} tests[] = {
    { nxt_string("APISID=HHIHYY4UEPWGuLbl/AmPSNDWCS75_0DQgA"),
      nxt_string("APISID"),
      nxt_string("HHIHYY4UEPWGuLbl/AmPSNDWCS75_0DQgA") },

    { nxt_string("NID=511=r_iLIzzHAVM27rpFL"),
      nxt_string("NID"),
      nxt_string("511=r_iLIzzHAVM27rpFL") },

    { nxt_string("=511=r_iLIzzHAVM27rpFL; SEARCH_SAMESITE=CgQIxJQB"),
      nxt_string("SEARCH_SAMESITE"),
      nxt_string("CgQIxJQB") },

    { nxt_string("NID=;"),
      nxt_string("NID"),
      nxt_string("") },

    { nxt_string("NID;"),
      nxt_string("NID"),
      nxt_null_string },

    { nxt_string("NID="),
      nxt_string("NID"),
      nxt_string("") },

    { nxt_string("NID"),
      nxt_string("NID"),
      nxt_null_string },
};`

**Comment by changxiaocui** [comment]
After trying, I decided to add descriptions instead of adding tests.

As per https://datatracker.ietf.org/doc/html/rfc6265#section-4.1.1
cookie-value = *cookie-octet / ( DQUOTE *cookie-octet DQUOTE )
cookie-octet = %x21 / %x23-2B / %x2D-3A / %x3C-5B / %x5D-7E

So '=' is a valid character in cookie-value, but the previous parser ignored this.
The code of the new parser is clear enough to read.

Below is some code when I tried to add tests. Most of the examples are from
the google search client.
`    static struct {
        nxt_str_t  cookie;
        nxt_str_t  name;
        nxt_str_t  value;

    } tests[] = {
        { nxt_string("APISID=HHIHYY4UEPWGuLbl/AmPSNDWCS75_0DQgA"),
          nxt_string("APISID"),
          nxt_string("HHIHYY4UEPWGuLbl/AmPSNDWCS75_0DQgA") },

        { nxt_string("NID=511=r_iLIzzHAVM27rpFL"),
          nxt_string("NID"),
          nxt_string("511=r_iLIzzHAVM27rpFL") },

        { nxt_string("=511=r_iLIzzHAVM27rpFL; SEARCH_SAMESITE=CgQIxJQB"),
          nxt_string("SEARCH_SAMESITE"),
          nxt_string("CgQIxJQB") },

        { nxt_string("NID=;"),
          nxt_string("NID"),
          nxt_string("") },

        { nxt_string("NID;"),
          nxt_string("NID"),
          nxt_null_string },

        { nxt_string("NID="),
          nxt_string("NID"),
          nxt_string("") },

        { nxt_string("NID"),
          nxt_string("NID"),
          nxt_null_string },
    };`

**Comment by hongzhidao** [comment]
@changxiaocui 
Thank you, will review it.

**Comment by hongzhidao** [comment]
> > In fact, It's hard to fix these issues without refactoring.
> 
> Maybe you can refactor without fixing the issues? So you could refactor as the first commit; and then fix the issue.

@changxiaocui Are you on it?
I agree it's better to split it into two commits, one is for refactoring, the other is for fixing.

**Comment by changxiaocui** [comment]
I am adding tests.

File: src/nxt_http_request.c [hunk 1]
```diff
@@ -947,44 +947,70 @@ nxt_http_cookies_parse(nxt_http_request_t *r)
...
+        default:
+            break;
+        }
     }
 
-    if (name != NULL) {
-        nv = nxt_http_cookie(cookies, name, name_length, start, p);
-        if (nxt_slow_path(nv == NULL)) {
-            return NXT_ERROR;
+    if (start < end) {
```
Comment (changxiaocui) [inline src/nxt_http_request.c:1000]: if (name.length || start < end)

**Comment by hongzhidao** [comment]
Hi, @changxiaocui 

I just found I ever did a similar fixing in arguments parsing, it's more concise.
https://github.com/nginx/unit/commit/40ad333a9c938f73621af0f32b0f09ab634c26c0

Would you like to refer to it?

**Comment by changxiaocui** [comment]
> 

Thanks, I'll take a look. I've done extensive testing, both functional and benchmarking, against multiple implementations. I need some time to analyze and summarize.

**Comment by changxiaocui** [comment]
> Hi, @changxiaocui
> 
> I just found I ever did a similar fixing in arguments parsing, it's more concise. [40ad333](https://github.com/nginx/unit/commit/40ad333a9c938f73621af0f32b0f09ab634c26c0)
> 
> Would you like to refer to it?

Thanks, I'll take a look. I've done extensive testing, both functional and benchmarking, against multiple implementations. I need some time to analyze and summarize.

**Comment by hongzhidao** [comment]
That’s great, I suggest you look into the implementation of arguments parsing, they are quite similar, but remember not to refactor out more generic. If it makes sense to do such refactoring, they can be postponed. But I guess it doesn’t mean much.

**Comment by changxiaocui** [comment]
> That’s great, I suggest you look into the implementation of arguments parsing, they are quite similar, but remember not to refactor out more generic. If it makes sense to do such refactoring, they can be postponed. But I guess it doesn’t mean much.

It is perfect now.

**Comment by changxiaocui** [comment]
`#include <stdio.h>
#include <string.h>
#include <stddef.h>
#include <time.h>


typedef unsigned char u_char;

typedef struct {
    size_t      len;
    u_char     *data;
} ngx_str_t;

typedef struct {
    ngx_str_t  name;
    ngx_str_t  value;
} ngx_pair_t;


#define ngx_length(s)                                                        \
    (sizeof(s) - 1)
#define ngx_string(str)     { ngx_length(str), (u_char *) str }
#define ngx_null_string     { 0, NULL }
#define ngx_str_set(str, text)                                               \
    (str)->len = sizeof(text) - 1; (str)->data = (u_char *) text
#define ngx_str_null(str)   (str)->len = 0; (str)->data = NULL

#define ngx_strchr(s1, c)   strchr((const char *) s1, (int) c)
#define ngx_memchr(s, c, length)                                              \
    memchr((char *) s, c, length)


static int
ngx_http_proxy_parse_cookie_old(ngx_str_t *value, ngx_pair_t *pair)
{
    u_char        *start, *end, *p, *last;
    ngx_str_t      name, val;

    ngx_str_null(&pair->name);
    ngx_str_null(&pair->value);

    start = value->data;
    end = value->data + value->len;

    for ( ;; ) {

#if 1
        last = (u_char *) ngx_strchr(start, ';');
#else
        for (last = start; last < end; last++) {
            if (*last == ';') { break; }
        }
#endif

        if (last == NULL) {
            last = end;
        }

        while (start < last && *start == ' ') { start++; }

        for (p = start; p < last && *p != '='; p++) { /* void */ }

        name.data = start;
        name.len = p - start;

        while (name.len && name.data[name.len - 1] == ' ') {
            name.len--;
        }

        if (p < last) {

            p++;

            while (p < last && *p == ' ') { p++; }

            val.data = p;
            val.len = last - val.data;

            while (val.len && val.data[val.len - 1] == ' ') {
                val.len--;
            }

        } else {
            ngx_str_null(&val);
        }

        ngx_str_null(&pair->name);
        ngx_str_null(&pair->value);

	    pair->name = name;
	    pair->value = val;
#if 1
        start = last + 1;
        if (start >= end) {
            break;
        }
#else
        if (last == end) {
            break;
        }

        start = last + 1;
#endif
    }

    return 0;
}

static int
ngx_http_proxy_parse_cookie_new(ngx_str_t *value, ngx_pair_t *pair)
{
    u_char        *start, *end, *p;
    ngx_str_t      name, val, *nv;

    ngx_str_null(&pair->name);
    ngx_str_null(&pair->value);

    ngx_str_null(&val);
    ngx_str_null(&name);
    start = value->data;
    end = value->data + value->len;

    for (p = start; p < end; p++) {
        switch (*p) {
        case '=':
            if (name.data != NULL) { break; }

            while (*start == ' ') { start++; }
            name.data = start;
            name.len = p - start;
            while (name.len && name.data[name.len - 1] == ' ') { name.len--; }
            //while (name.data[name.len - 1] == ' ' && name.len) { name.len--; }

            start = p + 1;
            break;

        case ';':
            nv = name.data == NULL ? &name : &val;

            while (*start == ' ') { start++; }
            nv->data = start;
            nv->len = p - start;
            while (nv->len && nv->data[nv->len - 1] == ' ') { nv->len--; }
            //while (nv->data[nv->len - 1] == ' ' && nv->len) { nv->len--; }

            ngx_str_null(&pair->name);
            ngx_str_null(&pair->value);

	        pair->name = name;
	        pair->value = val;

            ngx_str_null(&name);
            ngx_str_null(&val);

            start = p + 1;
            break;
        }
    }

    if (name.len || start < end) {
        nv = name.data == NULL ? &name : &val;

        while (*start == ' ' && start < p) { start++; }
        nv->data = start;
        nv->len = p - start;
        while (nv->len && nv->data[nv->len - 1] == ' ') { nv->len--; }
        //while (nv->data[nv->len - 1] == ' ' && nv->len) { nv->len--; }

        ngx_str_null(&pair->name);
        ngx_str_null(&pair->value);

	    pair->name = name;
	    pair->value = val;
    }

    return 0;
}

static int
ngx_http_cookie_parse_old(ngx_str_t *cookie, ngx_pair_t *pair)
{
    u_char                 *p, *start, *end;
    ngx_str_t              name, value, *pnv;

    ngx_str_null(&pair->name);
    ngx_str_null(&pair->value);

    ngx_str_null(&name);
    ngx_str_null(&value);
    start = cookie->data;
    end = start + cookie->len;

    for (p = start; p < end; p++) {
        switch (*p) {
        case '=':
            if (name.data != NULL) { break; }

            while (start[0] == ' ') { start++; }
            name.data = start;
            name.len = p - start;
            while (name.len && name.data[name.len - 1] == ' ') { name.len--; }

            start = p + 1;
            break;

        case ';':
            pnv = name.data == NULL ? &name : &value;

            while (start[0] == ' ') { start++; }
            pnv->data = start;
            pnv->len = p - start;
            while (pnv->len && pnv->data[pnv->len - 1] == ' ') { pnv->len--; }

            if (name.len > 0) {
                ngx_str_null(&pair->name);
                ngx_str_null(&pair->value);

	            pair->name = name;
	            pair->value = value;
            }

            ngx_str_null(&name);
            ngx_str_null(&value);

            start = p + 1;
            break;
        }
    }

    if (name.len || start < end) {
        pnv = name.data == NULL ? &name : &value;

        while (start[0] == ' ' && start < end) { start++; }
        pnv->data = start;

        while (end[-1] == ' ' && end > start) { end--; }
        pnv->len = end - start;

        if (name.len > 0) {
            ngx_str_null(&pair->name);
            ngx_str_null(&pair->value);

	        pair->name = name;
	        pair->value = value;
        }
    }

    return 0;
}

static int
ngx_http_cookie_parse_new(ngx_str_t *value, ngx_pair_t *pair)
{
    u_char        *start, *end, *p, *last;
    ngx_str_t      name, val;

    ngx_str_null(&pair->name);
    ngx_str_null(&pair->value);

    start = value->data;
    end = value->data + value->len;

    while (start < end) {

        last = (u_char *) ngx_strchr(start, ';');

        if (last == NULL) {
            last = end;
        }

        while (start < last && *start == ' ') { start++; }

	    p = (u_char *) ngx_memchr(start, '=', (size_t)(last - start));

        if (p == NULL) {
            p = last;
        }

        name.data = start;
        name.len = p - start;

        while (name.len && name.data[name.len - 1] == ' ') {
            name.len--;
        }

        if (p < last) {

            p++;

            while (p < last && *p == ' ') { p++; }

            val.data = p;
            val.len = last - val.data;

            while (val.len && val.data[val.len - 1] == ' ') {
                val.len--;
            }

        } else {
            ngx_str_null(&val);
        }

        ngx_str_null(&pair->name);
        ngx_str_null(&pair->value);

	    pair->name = name;
	    pair->value = val;

        start = last + 1;
    }

    return 0;
}

static int
ngx_http_cookie_parse_test(ngx_str_t *cookie, ngx_pair_t *pair)
{
    size_t                 name_length;
    u_char                 *p, *last, *name, *value, *start, *end;

    ngx_str_null(&pair->name);
    ngx_str_null(&pair->value);

    start = cookie->data;
    end = cookie->data + cookie->len;

    while (start < end) {
        last = (u_char *) ngx_memchr(start, ';', end - start);
        if (last == NULL) {
            last = end;
        }

        while (start < last && start[0] == ' ') { start++; }

        value = (u_char *) ngx_memchr(start, '=', last - start);
        if (value == NULL) {
            value = last;
        }

        p = value;
        while (p > start && p[-1] == ' ') { p--; }

        name_length = p - start;
        name = start;

        p = last;

        if (value < last) {
            value++;
            while (value < last && value[0] == ' ') { value++; }
            while (p > value && p[-1] == ' ') { p--; }
        } else {
            value = NULL;
        }

        if (name_length > 0) {
            ngx_str_null(&pair->name);
            ngx_str_null(&pair->value);

	        pair->name.data = name;
	        pair->name.len = name_length;
            pair->value.data = value;
            pair->value.len = value == NULL ? 0 : (p - value);
        }

        start = last + 1;
    }
}

static int
ngx_http_cookie_parse_final(ngx_str_t *cookie, ngx_pair_t *pair)
{
    size_t                 name_length;
    u_char                 *p, *last, *name, *value, *end;

    ngx_str_null(&pair->name);
    ngx_str_null(&pair->value);

    end = cookie->data + cookie->len;

    for (name = cookie->data; name < end; name = last + 1) {

        last = ngx_memchr(name, ';', end - name);
        if (last == NULL) {
            last = end;
        }

        while (name < last && name[0] == ' ') { name++; }

        value = ngx_memchr(name, '=', last - name);
        if (value == NULL) {
            value = last;
        }

        for (p = value; p > name && p[-1] == ' '; p--) { /* void */ }

        name_length = p - name;

        if (name_length > 0) {
            p = last;

            if (value < last) {
                value++;
                while (value < last && value[0] == ' ') { value++; }
                while (p > value && p[-1] == ' ') { p--; }
            } else {
                value = NULL;
            }

            ngx_str_null(&pair->name);
            ngx_str_null(&pair->value);

	        pair->name.data = name;
	        pair->name.len = name_length;
            pair->value.data = value;
            pair->value.len = value == NULL ? 0 : (p - value);
        }
    }
}

typedef struct {
    ngx_str_t  cookie;
    ngx_str_t  name;
    ngx_str_t  value;
} ngx_tests_t;

typedef int (*ngx_cookie_parse_handler_pt)(ngx_str_t *value, ngx_pair_t *pair);

static int ngx_test_once(ngx_tests_t *tests, int max, ngx_cookie_parse_handler_pt parser);
static int ngx_bench_once(ngx_tests_t *tests, int max, ngx_cookie_parse_handler_pt parser);

int main() {
    static ngx_tests_t tests[] = {  // 2.8 vs 6.45
        { ngx_string("APISID=HHIHYY4UEPWGuLbl/AmPSNDWCS75_0DQgA"),  // 0.34 vs 1.37
          ngx_string("APISID"),
          ngx_string("HHIHYY4UEPWGuLbl/AmPSNDWCS75_0DQgA") },

        { ngx_string("  APISID  =  HHIHYY4UEPWGuLbl/AmPSNDWCS75_0DQgA    ;"),  // 0.58 vs 1.82
          ngx_string("APISID"),
          ngx_string("HHIHYY4UEPWGuLbl/AmPSNDWCS75_0DQgA") },

        { ngx_string("NID=511=r_iLIzzHAVM27rpFL"),  // 0.26 vs 0.77
          ngx_string("NID"),
          ngx_string("511=r_iLIzzHAVM27rpFL") },

        { ngx_string("=511=r_iLIzzHAVM27rpFL; SEARCH_SAMESITE=CgQIxJQB"),  // 0.72 vs 1.58
          ngx_string("SEARCH_SAMESITE"),
          ngx_string("CgQIxJQB") },

        { ngx_string("NID=;"),  // == 0.27 vs 0.27
          ngx_string("NID"),
          ngx_string("") },

        { ngx_string("NID;"),  // 0.23 vs 0.24
          ngx_string("NID"),
          ngx_null_string },
/*
*/

        { ngx_string("NID="),  // 0.28 vs 0.26
          ngx_string("NID"),
          ngx_string("") },

        { ngx_string("NID"),  // 0.22 vs 0.21
          ngx_string("NID"),
          ngx_null_string },
/*
*/
    };

    printf("Old parser start:\n");
    if (ngx_test_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_proxy_parse_cookie_old)) {
        printf("Old parser: failed!!!\n");
        return -1;
    }

    int i = 0;
    clock_t start;
    static int bench_max = 10000000;

    start = clock();
    for (i = 0; i < bench_max; i++) {
        ngx_bench_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_proxy_parse_cookie_old);
    }
    printf("Old parser consuming: %f seconds\n", (double)(clock() - start)/CLOCKS_PER_SEC);



    printf("New parser start:\n");
    if (ngx_test_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_proxy_parse_cookie_new)) {
        printf("New parser: failed!!!\n");
        return -2;
    }

    start = clock();
    for (i = 0; i < bench_max; i++) {
        ngx_bench_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_proxy_parse_cookie_new);
    }
    printf("New parser consuming: %f seconds\n", (double)(clock() - start)/CLOCKS_PER_SEC);



    printf("Unit parser start:\n");
    if (ngx_test_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_cookie_parse_old)) {
        printf("Unit parser: failed!!!\n");
        return -3;
    }

    start = clock();
    for (i = 0; i < bench_max; i++) {
        ngx_bench_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_cookie_parse_old);
    }
    printf("Unit parser consuming: %f seconds\n", (double)(clock() - start)/CLOCKS_PER_SEC);



    printf("Test parser start:\n");
    if (ngx_test_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_cookie_parse_test)) {
        printf("Test parser: failed!!!\n");
        return -4;
    }

    start = clock();
    for (i = 0; i < bench_max; i++) {
        ngx_bench_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_cookie_parse_test);
    }
    printf("Test parser consuming: %f seconds\n", (double)(clock() - start)/CLOCKS_PER_SEC);



    printf("Final parser start:\n");
    if (ngx_test_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_cookie_parse_final)) {
        printf("Final parser: failed!!!\n");
        return -4;
    }

    start = clock();
    for (i = 0; i < bench_max; i++) {
        ngx_bench_once(tests, sizeof(tests)/sizeof(ngx_tests_t), ngx_http_cookie_parse_final);
    }
    printf("Final parser consuming: %f seconds\n", (double)(clock() - start)/CLOCKS_PER_SEC);


    return 0;
}

#define ngx_memcmp(s1, s2, n)  memcmp((const char *) s1, (const char *) s2, n)
#define ngx_str_eq(s1, s2)                                                     \
    (((s1)->data == NULL) && ((s2)->data == NULL)                              \
     || ((s1)->len == (s2)->len) && (ngx_memcmp((s1)->data, (s2)->data, (s1)->len) == 0))

static int
ngx_test_once(ngx_tests_t *tests, int max, ngx_cookie_parse_handler_pt parser)
{
    static ngx_pair_t result = ngx_null_string;
    int i = 0;

    for (i = 0; i < max; i++) {

	    parser(&tests[i].cookie, &result);

#if 0
        printf("COOKIE [%.*s];\nexpect [%.*s] [%.*s];\ngot [%.*s] [%.*s]\n",
               tests[i].cookie.len, tests[i].cookie.data, tests[i].name.len, tests[i].name.data,
               tests[i].value.len, tests[i].value.data, result.name.len, result.name.data,
               result.value.len, result.value.data);
#endif

	    if (!ngx_str_eq(&result.name, &tests[i].name)) {
            printf("Got pair name [%.*s], but expect [%.*s] while cookie is [%.*s]\n",
                   result.name.len, result.name.data, tests[i].name.len,
                   tests[i].name.data, tests[i].cookie.len, tests[i].cookie.data);
            return -1;
        }

        if (!ngx_str_eq(&result.value, &tests[i].value)) {
            printf("Got pair value [%.*s], but expect [%.*s] while cookie is [%.*s]\n",
                   result.value.len, result.value.data, tests[i].value.len,
                   tests[i].value.data, tests[i].cookie.len, tests[i].cookie.data);
            return -2;
        }
    }

    return 0;
}

static int
ngx_bench_once(ngx_tests_t *tests, int max, ngx_cookie_parse_handler_pt parser)
{
    static ngx_pair_t result = ngx_null_string;
    int i = 0;

    for (i = 0; i < max; i++) {
	    parser(&tests[i].cookie, &result);
    }

    return 0;
}`

**Comment by ac000** [comment]
A couple of general comments

1) Could you rebase against current master to get rid of the merge commit for easier review.
2) Could you provide meaningful commit messages/summaries? Otherwise we don't know what the patch is _supposed_ to be doing and why...

Thanks.

**Review by ac000 — COMMENTED** [review]
EDIT: Not sure what happened but this comment is supposed to be in relation to https://github.com/nginx/unit/pull/711/commits/6a891f6a2fc17ca538281268fc261be6570b9b43

Is this just fixing an issue introduced by one of your other commits?

The code in question doesn't exist in unit master and seems to have been introduced by your merge commit.

**Pull request closed without merging**